Gateway: Add Cache-Control/max-age and ETag to IPNS requests

ion1 opened the pull request:
Based on https://github.com/ipfs/go-ipfs/pull/1887 “cache ipns entries to speed things up a little”.

Addresses https://github.com/ipfs/go-ipfs/issues/1818.

This is my first time learning Go, I hope I’m doing things somewhat correctly.

jbenet commented on core/corehttp/gateway_handler.go:
```diff
@@ -148,13 +154,11 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
-	// and only if it's /ipfs!
-	// TODO: break this out when we split /ipfs /ipns routes.
+	w.Header().Set("ETag", etag)
+	w.Header().Set("Cache-Control", fmt.Sprintf("public, max-age=%d", int(res.TTL.Seconds())))
```
nope, errors still happen below. this is trickty! --  maybe the error functions should _clear_ this header to be sure.

jbenet replied:
ensure it's `/ipfs/` route. code evolves!

ion1 commented:
* Added code to unset Cache-Control and ETag in the `webError` functions.
* Checked that all error responses in `getOrHeadHandler` go through those functions.
* Added a comment reminding to unset the headers in error responses if not using the functions.

jbenet commented:
Ok, this LGTM to me. @ion1 last thing is we should test some of this using sharness tests, so that the api  tests are carried to other impls eventually.

ion1 commented:
* Added sharness tests.
* Noticed that we weren’t following the RFC for ETags. Quote marks around the tag are mandatory. Fixed.

The sharness tests involving IPNS were `test_expect_failure` because `ipfs name publish` fails with `Error: failed to find any peer in table`.

All the new IPNS tests in the PR are `test_expect_success`es at the moment, and I verified them to succeed in a modified environment where they have access to the swarm.

Should I leave them as they are (which leaves the test suite failing until the `publish` problem is fixed) or change them all to `test_expect_failure`s?

ion1 commented:
I created https://github.com/ipfs/go-ipfs/issues/1941, changed them to `test_expect_failure`s and added a comment with a link to the issue.

ion1 commented:
I added sharness tests IPNS caching (#1934).

whyrusleeping commented on core/corehttp/gateway_test.go:
```diff
@@ -20,18 +21,33 @@ import (
+	return res.Path, err
+}
+
+func (m mockNamesys) ResolveFull(ctx context.Context, name string) (namesys.ResolveResult, error) {
```
I'm not sure if `ResolveFull` is the correct name for this.

ion1 replied:
You are probably right. Any ideas? `ResolveDetailed` perhaps?

whyrusleeping commented on core/corehttp/gateway_test.go:
```diff
@@ -114,21 +130,27 @@ func TestGatewayGet(t *testing.T) {
 		host   string
 		path   string
 		status int
+		ttl    []time.Duration // an approximation for a Maybe/Option sum type
```
you could just use `ttl *time.Duration` and pass `&kTTL` as the value.

ion1 replied:
I tried that first, but `&namesys.ImmutableTTL` failed with “cannot take the address of namesys.ImmutableTTL” IIRC.

whyrusleeping replied:
right, but thats different than `&kTTL`

ion1 replied:
The entries in the table being generated here here need to include “no TTL”, `namesys.ImmutableTTL` and `kTTL` as values for this field. I could create a local copy of `namesys.ImmutableTTL` to be able to use a pointer here, but would that be any cleaner?

whyrusleeping replied:
the clean way around that is to make `namesys.ImmutableTTL` into a `var` instead of a `const`

ion1 commented:
@whyrusleeping, thanks for your comments. I renamed `ResolveFull` as `ResolveDetailed` and made the test use a pointer to encode the possible lack of a TTL value. Does this look okay?

whyrusleeping commented on core/corehttp/gateway_test.go:
```diff
@@ -20,18 +21,33 @@ import (
+		return namesys.ResolveResult{}, namesys.ErrResolveFailed
 	}
-	return p, nil
+	return namesys.ResolveResult{entry.path, entry.ttl}, nil
```
use explicit naming here for struct fields.

whyrusleeping commented on core/corehttp/gateway_test.go:
```diff
@@ -20,18 +21,33 @@ import (
+	return m.ResolveNDetailed(ctx, name, namesys.DefaultDepthLimit)
+}
+
+func (m mockNamesys) ResolveNDetailed(ctx context.Context, name string, depth int) (namesys.ResolveResult, error) {
```
should probably return `*namesys.ResolveResult`, cleaner code in error cases

ion1 replied:
The TTL field is actually useful in error cases. Many errors should not be cached (such as when we fail to fetch something from the network but it could be available a second later) but some errors could certainly be cached:

* `/ipfs/QmFoo/file` is requested; `/ipfs/QmFoo` resolves correctly (TTL=“infinite”) but does not contain `file`: the error’s TTL can be “infinite”.
* `/ipns/QmBar/file` is requested; `/ipns/QmBar` resolves to `/ipfs/QmFoo` (TTL=2 minutes for example) but that does not contain `file`: the error’s TTL=2 minute because that’s how long the resolved IPNS entry can be cached.
* A parse error in the path being resolved: the error’s TTL can be “infinite”. e.g. `namesys/namesys.go`:

```go
        segments := strings.SplitN(name, "/", 3)
        if len(segments) < 3 || segments[0] != "" {
                log.Warningf("Invalid name syntax for %s", name)
                return ResolveResult{TTL: ImmutableTTL}, ErrResolveFailed
```

Nothing in the code caches errors or serves HTTP errors with a `max-age` *yet* but the infrastructure is there if we choose to implement caching for some errors later.

Would it be a reasonable solution to return `ttl time.Duration, result *ResolveResult, err error` where `ttl` is always given (usually 0 for errors but not always) and exactly one of `result`, `err` is nil?

whyrusleeping commented on core/corehttp/gateway_test.go:
```diff
@@ -114,21 +130,28 @@ func TestGatewayGet(t *testing.T) {
-	ns["/ipns/example.com"] = path.FromString("/ipfs/" + k)
+	kTTL := 10 * time.Minute
+	ns["/ipns/example.com"] = mockEntry{
+		path: path.FromString("/ipfs/" + k),
```
completely out of scope of this PR, but I'm thinking we should probably make `FromString` do the parsing check in `path.ParsePath`, and panic if it fails.

whyrusleeping commented on namesys/routing.go:
```diff
@@ -26,14 +26,14 @@ type routingResolver struct {
 }
 
-func (r *routingResolver) cacheGet(name string) (path.Path, bool) {
+func (r *routingResolver) cacheGet(name string) (ResolveResult, bool) {
```
probably want to return `*ResolveResult`, cleaner error cases.

whyrusleeping commented on namesys/routing.go:
```diff
@@ -156,29 +154,68 @@ func (r *routingResolver) resolveOnce(ctx context.Context, name string) (path.Pa
+	}
+
+	if cacheTil.Before(now) {
+		// Already EOL, do not cache.  This is unexpected, an expired
```
maybe log a warning?

whyrusleeping commented on test/sharness/t0100-name.sh:
```diff
@@ -10,12 +10,14 @@ test_description="Test ipfs repo operations"
 	PEERID=`ipfs id --format="<id>"` &&
 	test_check_peerid "${PEERID}" &&
-	ipfs name publish "/ipfs/$HASH_WELCOME_DOCS" >publish_out
+	ipfs name publish${TTL_PARAMS:+ $TTL_PARAMS} "/ipfs/$HASH_WELCOME_DOCS" >publish_out
```
this syntax is strange to me, afaict its saying:

If `TTL_PARAMS` is null or unset, use the value of the variable named by `TTL_PARAMS`.

ref: http://www.gnu.org/software/bash/manual/bash.html#Shell-Parameter-Expansion

ion1 replied:
If `TTL_PARAMS` is *set*, add a space followed by `$TTL_PARAMS`, otherwise add nothing.

whyrusleeping replied:
ah, misread that. Why add that complication over just putting the variable there?
```
ipfs name publish $TTL_PARAMS "/ipfs/$HASH_WELCOME_DOCS" >publish_out
```

ion1 replied:
I used the `:+` expansion in the test title to avoid a string with an extra space like `"'ipfs name publish ' succeeds"` when there is no ttl parameter, and then just used the exact same thing in the actual code. But granted, it’s much more readable that way. How about I keep the `:+` expansion in the title to make the output pretty but use a plain `$TTL_PARAMS` in the code?

whyrusleeping replied:
yeah, that works for me. Big priority with tests is for people to be able to jump right in and change/add things easily

whyrusleeping commented on test/sharness/t0110-gateway.sh:
```diff
@@ -22,58 +22,315 @@ apiport=$PORT_API
+
+ITEM="GET IPFS directory path"
+
+test_expect_success "$ITEM: succeeds" '
```
are all these tests with `$ITEM` in the description exactly the same? maybe make them into a function to clean things up a bit

ion1 replied:
They are different. This one checks for `test_should_contain "Index of /ipfs/$HASH2" actual` while the following one checks `test_cmp "$DATA2"/test actual`. The one after that checks that a redirect happens. The one after that checks a path which contains `index.html` has to check the ETag header against the path that was requested, not against the hash of the `index.html`.

It gets even more difficult to refactor them to a function later where we want to do an IPNS publish for the following item but check that the cache still returns results from the earlier IPNS publish for the current item.

This is an annoying case where all the items are just slightly different in various different ways and sh doesn’t exactly provide powerful ways to abstract things.

But I’ll give it another look and see if I can find a way to make it a bit better.

whyrusleeping replied:
ah yeah, i see now. that is rather annoying... yay shell!

whyrusleeping commented on test/sharness/t0130-multinode.sh:
```diff
@@ -53,6 +53,18 @@ check_file_fetch() {
 }
 
+test_resolve() {
+	_nid="$1"; shift
```
i would opt for `$1`, `$2`, `$3`. I think thats cleaner than shifting just to be able to use the same variable.

@chriscool what do you think?

ion1 replied:
This pattern is useful whenever you want to have a variable number of parameters in the end. You can take each mandatory parameter with `foo="$1"; shift` and you have all the remaining parameters (zero or more) in `"$@"` afterwards. I used that in a couple of the other tests, so I used the same style for consistency despite not using `"$@"` in this specific function.

In scripts that use `set -e`, the `shift` also acts as an additional sanity check that the mandatory parameter is actually there. On the other hand, with `set -u` just the `"$1"` expansion covers that. But this is irrelevant here as sharness suites do not use either.

chriscool replied:
When one want to use "$@", it's ok to use `shift`, so that the first arguments are not part of "$@". Here I think it would be better to use `$1`, `$2` and `$3` as there are other functions in the same file using `$1`, `$2` and `$3` already.

ion1 commented:
Thanks for your input, @whyrusleeping and @chriscool.

Done:

* Get rid of the `ResolveResult` struct, return `(Path, Duration, error)` instead. (The `Duration` applies to both success and error results.)
* Log a warning in the EOL sanity check in routing.go.
* Get rid of noisy `${TTL_PARAMS:+ $TTL_PARAMS}` where ` $TTL_PARAMS` is sufficient.
* Get rid of `param="$1"; shift` pattern where `"$@"` is not used.
* Refactor test/sharness/t0110-gateway.sh.

Not addressed:

> completely out of scope of this PR, but I'm thinking we should probably make `FromString` do the parsing check in `path.ParsePath`, and panic if it fails.

jbenet commented on core/corehttp/gateway_test.go:
```diff
@@ -20,18 +21,33 @@ import (
+	return p, err
+}
+
+func (m mockNamesys) ResolveDetailed(ctx context.Context, name string) (path.Path, time.Duration, error) {
```
dont like this name. not clear enough.

other suggestions:
```
ResolveWithTTL
ResolveWithCaching
```
hmmm not sure.

ion1 replied:
`ResolveWithTTL` sounds good. `ResolveWithCaching` might be ambiguous since this variant of `Resolve` does not affect whether caching occurs at all.

jbenet commented on core/corehttp/gateway_test.go:
```diff
@@ -20,18 +21,33 @@ import (
 
-func (m mockNamesys) ResolveN(ctx context.Context, name string, depth int) (value path.Path, err error) {
-	p, ok := m[name]
+func (m mockNamesys) Resolve(ctx context.Context, name string) (path.Path, error) {
```
i think keep the naming on these, it's for ease of reading on the api docs

jbenet commented on test/sharness/t0110-gateway.sh:
```diff
@@ -22,60 +22,197 @@ apiport=$PORT_API
+      '
+      ;;
+  esac
+}
```
this is _very_ hard to read. i think this can be way cleaner.

jbenet commented on test/sharness/t0110-gateway.sh:
```diff
@@ -22,60 +22,197 @@ apiport=$PORT_API
+    *)
+      # The go http package replaces ETag with Etag.  Accept both.  (HTTP
+      # header names are case-insensitive but ETag is the official name.)
+      "$TEST_COMMAND" "$_item: has ETag" '
```
dont do this, this leads to macro hell fast. just test both, write both out. it's much clearer and waaay easier to instrument + debug.

jbenet commented on test/sharness/t0160-resolve.sh:
```diff
@@ -14,104 +14,141 @@ test_expect_success "resolve: prepare files" '
+test_resolve "/ipns/$id_hash" "/ipfs/$c_hash"
 
-# should work online
-test_launch_ipfs_daemon
```
the benefit of doing it this way it's that it's the _exact_ same set of commands for both, so test _everything_ works

jbenet replied:
i dont understand why this is changing? just keep it in the function. if anything we should be adding the lines to test it offline

ion1 replied:
The diff algorithm just doesn’t realize it’s still in the file, please see line 87.

jbenet replied:
No, i understand it's there. The point is that the lines 89-138 should be in a function so they can be called with both the daemon online and offline, like several other tests. please keep the pattern:

```sh
test_launch_ipfs_daemon
test_some_function
test_kill_ipfs_daemon
```

ion1 replied:
Those lines explicitly test the caching done by the daemon in online mode. I can put them into a function but they will fail if executed while offline because there is no daemon to cache resolve results.

jbenet commented:
@ion1 Looking much better! some comments above.

ion1 commented:
@jbenet, thanks for your input.

Done:
* Make sure the exact same commands are tested while offline and while online whereever possible in t0160-resolve.sh.
  * Added `test_expect_success_1941` to test-lib.sh as a workaround to #1941 to facilitate using the same test code for the offline and the online cases. Once #1941 is fixed, `test_expect_success_1941` can be globally replaced with just `test_expect_success`.
* Rename `ResolveDetailed` as `ResolveWithTTL`.
* Retain the named return values in gateway_test.go.
* Make `test_item` more readable (I hope) in t0110-gateway.sh.
* Do not use `E[Tt]ag`, just do a case-insensitive match.

ion1 commented:
The Travis failure does not seem to have anything to do with my changes.

jbenet commented on core/corehttp/gateway_handler.go:
```diff
@@ -146,15 +154,14 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
+	// Remember to unset the Cache-Control/ETag headers before error
+	// responses!  The webError functions unset them automatically.
+	// TODO: Consider letting clients cache 404s.  Are there cases where
+	// that would hurt?
```
yes let's not cache 404s anytime soon.

ion1 replied:
It’s unclear to me where it would hurt, though. Given a successfully resolved `/ipfs/QmFoo`, the knowledge that “`/ipfs/QmFoo/fileA` exists” should be as cacheable as “`/ipfs/QmFoo/fileB` does not exist” for example. `QmFoo` is immutable after all.

The same applies to `/ipns/QmBar`. The record’s TTL dictates that the knowledge that “`/ipns/QmBar/fileA` exists” can be cached for just as long as that “`/ipns/QmBar/fileB` does not exist”.

whyrusleeping replied:
i think @ion1 has a point here, but its going to be difficult to do it *right*. A 404 should only be cached if its for a link under a valid resolved hash. A 'could not find this hash on the network' 404 should never be cached.

ion1 replied:
The code should already return a TTL of 0 for all network failures.

jbenet commented on core/corehttp/gateway_handler.go:
```diff
@@ -446,6 +453,7 @@ func webError(w http.ResponseWriter, message string, err error, defaultCode int)
 }
 
 func webErrorWithCode(w http.ResponseWriter, message string, err error, code int) {
+	unsetSuccessHeaders(w)
```
i really hope noone forgets to call this somewhere.

jbenet commented on namesys/interface.go:
```diff
@@ -47,6 +47,14 @@ const (
 	UnlimitedDepth = 0
+
+	// ImmutableTTL is the time-to-live value to be reported for immutable
+	// paths.
```
maybe note this is an HTTP thing.

jbenet replied:
actually, no, this should return a special value which is infinity -- immutable things are meant to be cached forever_. The http gateway can translate to the stupid http "forever means a year" thing.

ion1 replied:
Fair enough. It would be easy to create a data type like `data TTLDuration = Finite Duration | Infinity` in languages with custom sum types, and the ordering of the constructors would make `min`/`max` do the right thing, too. I wonder what the cleanest thing to do is in Go?

* Floats? (Eww)
* `interface {}` and dynamic typing? (Eww, too)
* A pointer to a Duration with nil signifying infinity?
* Something less ugly that I didn’t think of? (I am still new to Go.)

whyrusleeping replied:
how about -1 ?

jbenet commented on test/sharness/lib/test-lib.sh:
```diff
@@ -353,3 +355,38 @@ test_check_peerid() {
+# test_expect_success_1941 runs test_expect_failure if that is the case and
+# test_expect_success otherwise.  As soon as #1941 is fixed, all invocations of
+# test_expect_success_1941 can be replaced with test_expect_success and this
+# function can be removed.
```
idk how i feel about this. could hide problems. right now we just run things again. 

i'd be :-1: on this

ion1 replied:
I’m not sure what to do other than having #1941 block this PR. These tests will simply fail until it is fixed. Previously there were duplicated blocks of tests where the succeeding (offline) tests used `test_expect_success` and the failing (online) tests used `test_expect_failure` but that was not acceptable. Now both use the same `test_expect_success_1941` which is *temporary* and will be globally replaced with `test_expect_success` as soon as #1941 is fixed.

#1941 blocking this PR would be less than optimal because the code *works*, the testing environment just fails to do IPNS right. (And if you run the tests without the `ipfs bootstrap rm --all`, they all pass.)

jbenet replied:
> Previously there were duplicated blocks of tests where the succeeding (offline) tests used test_expect_success and the failing (online) tests used test_expect_failure but that was not acceptable. 

agreed.

> Now both use the same test_expect_success_1941 which is temporary and will be globally replaced with test_expect_success as soon as #1941 is fixed.

maybe #1941 should be fixed beforehand then. what will it take? cc @whyrusleeping 

@ion1 my reluctance is that once something is merged in, it often takes a very long time to remove it. For example, the writable gateway was working a long time ago. it was briefly temporarily removed, but that "temporary" lasted many months.

in this case, it may be ok as the code should do the right thing.

in any case, can this even merge on master? @whyrusleeping has asked for a freeze and for all new code to be added to dev0.4.0. If @whyrusleeping wants it to be on 0.4.0, then should

- [ ] rebase this on dev0.4.0
- [ ] see what it will take to fix #1941

jbenet commented on test/sharness/t0160-resolve.sh:
```diff
@@ -14,104 +14,129 @@ test_expect_success "resolve: prepare files" '
+	case "$src" in
+	  /ipns/*) test_cmd=test_expect_success_1941 ;;
+	  *)       test_cmd=test_expect_success ;;
+	esac
```
:-1:

jbenet commented:
@ion1  thanks -- more CR. and rebase to keep abreast of recent merges.

ion1 commented:
@jbenet, thanks for your input.

Done:
* Rebase.
* Add util/infduration with the constructors `InfiniteDuration()` and `FiniteDuration(d)`, use it in the TTL code.

I’m looking forward to your response to my comment about `test_expect_success_1941`. I’d like to emphasize that pending a fix for #1941, the tests running `ipfs name publish` while the daemon is running with zero peers will fail unconditionally.

jbenet commented on test/sharness/t0100-name.sh:
```diff
@@ -10,12 +10,14 @@ test_description="Test ipfs repo operations"
 
 test_init_ipfs
 
+for TTL_PARAMS in "" "--ttl=0s" "--ttl=1h"; do
```
wow the rest of the file is in a for loop without indentation. didnt even catch this the first time. please dont do that. 

do this instead:

```sh
somefunc() {
  ttl=$1
  all the tests here
}

for TTL in "" "--ttl=0s" "--ttl=1h"; do
  somefunc $TTL
done
```

whyrusleeping commented:
@ion1 since this PR touches a lot of files, I'd prefer if we put this on 0.4.0. Give rebasing it a shot, let me know how it goes.

whyrusleeping commented:
also, the infinite duration stuff (which i'm still not convinced is better than just using -1 as a special value) should go in thirdparty, i'm working on shrinking the util package.

whyrusleeping commented:
@ion1 update here?

ion1 commented:
@whyrusleeping: I have been waiting for a fix for #1941 which prevents the tests in this PR from working. I will rebase this and address the other remaining comments at that point.

whyrusleeping commented:
Has been idle for about a year. Closing for now. Please do feel free to reopen this as necessary. Closing old PRs just helps us more easily see what needs immediate attention
